cpp/src/arrow/dataset/file_base.cc

```diff
@@ -89,6 +89,14 @@ Result<std::shared_ptr<io::InputStream>> FileSource::OpenCompressed(
   return io::CompressedInputStream::Make(codec.get(), std::move(file));
 }
 
+bool FileSource::Equals(const FileSource& other) const {
+  bool match_file_system =
+      (filesystem_ == nullptr && other.filesystem_ == nullptr) ||
+      (filesystem_ && other.filesystem_ && filesystem_->Equals(other.filesystem_));
+  return file_info_.Equals(other.file_info_) && match_file_system &&
```

Comment:
nit, but put `match_file_system` first to avoid unnecessarily evaluating `file_info_.Equals`?

Reply:
ah yes good catch.